src/client.rs

```diff
@@ -236,69 +247,76 @@ where
-            }
-            TargetAddr::Ip(SocketAddr::V6(addr)) => {
-                debug!("TargetAddr::IpV6");
-                padding = 22;
-
-                packet[3] = 0x04;
-                debug!("addr ip {:?}", (*addr.ip()).octets());
-                packet[4..20].copy_from_slice(&(addr.ip()).octets()); // ip
-                packet[20..padding].copy_from_slice(&addr.port().to_be_bytes());
-                // port
-            }
-            TargetAddr::Domain(ref domain, port) => {
-                debug!("TargetAddr::Domain");
-                if domain.len() > u8::max_value() as usize {
-                    return Err(SocksError::ExceededMaxDomainLen(domain.len()));
+        packet[..3].copy_from_slice(&[consts::SOCKS5_VERSION, cmd.as_u8(), 0x00]);
+
+        match self.target_addr.as_ref() {
+            None => {
+                if cmd == Socks5Command::UDPAssociate {
+                    debug!("UDPAssociate without target_addr, fallback to zeros.");
+                    padding = 10;
+
+                    packet[3] = 0x01;
+                    packet[4..8].copy_from_slice(&[0, 0, 0, 0]); // ip
+                    packet[8..padding].copy_from_slice(&[0, 0]); // port
+                } else {
+                    return Err(anyhow::Error::msg("target addr should be present").into());
                 }
-                padding = 5 + domain.len() + 2;
-
-                packet[3] = 0x03; // Specify domain type
-                packet[4] = domain.len() as u8; // domain length
-                packet[5..(5 + domain.len())].copy_from_slice(domain.as_bytes()); // domain content
-                packet[(5 + domain.len())..padding].copy_from_slice(&port.to_be_bytes());
-                // port content (.to_be_bytes() convert from u16 to u8 type)
             }
+            Some(target_addr) => match target_addr {
+                TargetAddr::Ip(SocketAddr::V4(addr)) => {
+                    debug!("TargetAddr::IpV4");
+                    padding = 10;
+
+                    packet[3] = 0x01;
+                    debug!("addr ip {:?}", (*addr.ip()).octets());
+                    packet[4..8].copy_from_slice(&(addr.ip()).octets()); // ip
+                    packet[8..padding].copy_from_slice(&addr.port().to_be_bytes());
+                    // port
+                }
+                TargetAddr::Ip(SocketAddr::V6(addr)) => {
+                    debug!("TargetAddr::IpV6");
+                    padding = 22;
+
+                    packet[3] = 0x04;
+                    debug!("addr ip {:?}", (*addr.ip()).octets());
+                    packet[4..20].copy_from_slice(&(addr.ip()).octets()); // ip
+                    packet[20..padding].copy_from_slice(&addr.port().to_be_bytes());
+                    // port
+                }
+                TargetAddr::Domain(ref domain, port) => {
+                    debug!("TargetAddr::Domain");
+                    if domain.len() > u8::max_value() as usize {
```

Comment:
`u8::max_value()` is deprecated, could you replace it with MAX?

Reply:
Sure, I will fix that.